anospp_analysis/nn.py

```diff
@@ -0,0 +1,501 @@
+
+    #Read in exceptions from true_multi_targets file
+    for idx, item in true_multi_targets.iterrows():
+        potentially_affected_samples = comb_stats_df.loc[comb_stats_df[f'res_{item.level}'] == item.sgp, 'sample_id']
+        affected_samples = non_error_hap_df.query('sample_id in @potentially_affected_samples & target == @item.target') \
+            .groupby('sample_id').filter(lambda x: x['seqid'].nunique() > 2 & \
+                                         x['seqid'].nunique() < item.admissable_alleles)['sample_id'].unique()
+        comb_stats_df.loc[comb_stats_df.sample_id.isin(affected_samples), 'multiallelic_mosq_targets'] -= 1
+
+    comb_stats_df.loc[comb_stats_df.multiallelic_mosq_targets>int(ma_hi_threshold), 'contamination_risk'] = 'high'
+    comb_stats_df.loc[((comb_stats_df.multiallelic_mosq_targets>int(ma_med_threshold)) & \
+                       (comb_stats_df.multiallelic_mosq_targets<=int(ma_hi_threshold))) |\
+        (comb_stats_df.mosq_reads<int(rc_med_threshold)), 'contamination_risk'] = 'medium'
+    comb_stats_df.loc[comb_stats_df.contamination_risk.isnull(), 'contamination_risk'] = 'low'
+
+    logging.info(f"Identified {(comb_stats_df.contamination_risk=='high').sum()} samples with high contamination risk \
+        \n and {(comb_stats_df.contamination_risk=='medium').sum()} samples with medium contamination risk")
+
+    return(comb_stats_df)
+
+def generate_hard_calls(comb_stats_df, non_error_hap_df, test_samples, result_coarse, \
+                        result_int, result_fine, true_multi_targets, nn_asgn_threshold, \
+                        rc_med_threshold, ma_med_threshold, ma_hi_threshold):
+
+    logging.info('generating NN calls from assignment info')
+
+    #Account for filtering and error removal
+    comb_stats_df = recompute_sample_coverage(comb_stats_df, non_error_hap_df)
+
+    #Record whether NN assignment was performed
+    comb_stats_df.loc[comb_stats_df.sample_id.isin(test_samples), 'NN_assignment'] = 'yes'
+    comb_stats_df.loc[comb_stats_df.NN_assignment.isnull(), 'NN_assignment'] = 'no'
+
+    #Generate assignment hard calls if the threshold is met
+    for result, rescol in zip([result_coarse, result_int, result_fine], ['res_coarse', 'res_int', 'res_fine']):
+        asgn_dict = dict(result.loc[(result>=float(nn_asgn_threshold)).any(axis=1)].apply(\
+            lambda row: result.columns[row>=float(nn_asgn_threshold)][0], axis=1))
+        comb_stats_df[rescol] = comb_stats_df.sample_id.map(asgn_dict)
+
+    comb_stats_df = estimate_contamination(comb_stats_df, non_error_hap_df, true_multi_targets, \
+                                           rc_med_threshold, ma_med_threshold, ma_hi_threshold)
+
+    return(comb_stats_df)
+
+def plot_assignment_proportions(result, level, colors, nn_asgn_threshold):
+    
+    logging.info(f'Generate {level} level plots')
+    #Generate bar plots at given assignment level
+    fig, ax = plt.subplots(figsize=(20,7))
```

Comment:
suggest that plot width should depend on number of samples

Reply:
Done.

Reply:
Although it's a bit tricky with the legend - I force it to have 7 columns because there are so many species groups.